File: packages/sdk/src/commands/fork.ts
// fork command — WD-10. Locally fork a drop: clone the bundle source,
// rewrite SKILL.md (id, author, forked_from, version reset to 1.0.0).
//
// In v1 we accept a LOCAL source (directory or .zip) via --from. Once the
// registry ships, this command will also accept a registry drop-id and
// fetch the bundle automatically (and POST /drops/{id}/fork — WD-19).

import { cp, mkdir, readFile, writeFile } from "node:fs/promises";
import { existsSync, readdirSync } from "node:fs";
import { resolve } from "node:path";
import { parse as parseYaml, stringify as stringifyYaml } from "yaml";

import chalk from "../lib/chalk-shim.js";
import { loadCredential } from "../lib/eternitas.js";

export interface ForkOptions {
  from: string;          // local path to source drop directory
  newId: string;         // new drop id
  out: string;           // target directory for the fork
  newName?: string;      // optional human-readable name; defaults to source name + " (forked)"
  authorName?: string;   // overrides credentials/anonymous
  authorPassport?: string;
}

export interface ForkResult {
  sourceDir: string;
  targetDir: string;
  newId: string;
  forkedFrom: string;
}

const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---(\r?\n[\s\S]*)?$/;

export async function fork(opts: ForkOptions): Promise<ForkResult> {
  const sourceDir = resolve(opts.from);
  if (!existsSync(sourceDir)) {
    throw new Error(`source not found: ${sourceDir}`);
  }

  const target = resolve(opts.out);
  if (existsSync(target)) {
    const contents = readdirSync(target);
    if (contents.length > 0) {
      throw new Error(`target directory not empty: ${target}`);
    }
  } else {
    await mkdir(target, { recursive: true });
  }

  // Copy the source directory in full.
  await cp(sourceDir, target, { recursive: true });

  // Rewrite SKILL.md manifest.
  const skillPath = resolve(target, "SKILL.md");
  const raw = await readFile(skillPath, "utf-8");
  const match = raw.match(FRONTMATTER_RE);
  if (!match) {
    throw new Error(`source SKILL.md has no frontmatter: ${skillPath}`);
  }

  const manifest = parseYaml(match[1] ?? "") as Record<string, unknown>;
  if (!manifest || typeof manifest !== "object") {
    throw new Error(`source SKILL.md frontmatter must be a YAML object: ${skillPath}`);
  }

  const sourceId = String(manifest["id"] ?? "");
  const sourceName = manifest["name"];

  // Determine new author.
  let newAuthor: Array<Record<string, unknown>>;
  if (opts.authorName) {
    const author: Record<string, unknown> = { name: opts.authorName };
    if (opts.authorPassport) author["passport"] = opts.authorPassport;
    newAuthor = [author];
  } else {
    try {
      const cred = loadCredential({});
      newAuthor = [{
        name: cred.passport || "Your Name",
        ...(cred.passport ? { passport: cred.passport } : {}),
      }];
    } catch {
      newAuthor = [{ name: "Your Name" }];
    }
  }

  manifest["id"] = opts.newId;
  manifest["author"] = newAuthor;
  manifest["forked_from"] = sourceId;
  manifest["version"] = "1.0.0";
  if (opts.newName) {
    manifest["name"] = opts.newName;
  } else if (typeof sourceName === "string") {
    manifest["name"] = `${sourceName} (forked)`;
  }
  // Strip the source signature — a fork is unsigned until the new author signs.
  delete manifest["signature"];

  const yamlOut = stringifyYaml(manifest, { lineWidth: 0, indent: 2 });
  const body = match[2] ?? "";
  await writeFile(skillPath, `---\n${yamlOut.trimEnd()}\n---${body}`, "utf-8");

  return {
    sourceDir,
    targetDir: target,
    newId: opts.newId,
    forkedFrom: sourceId,
  };
}

export async function run(opts: ForkOptions): Promise<void> {
  try {
    const result = await fork(opts);
    process.stdout.write(
      `${chalk.green("✓")} forked ${result.forkedFrom} → ${result.newId}\n`,
    );
    process.stdout.write(`  at ${result.targetDir}\n\n`);
    process.stdout.write(`Next steps:\n`);
    process.stdout.write(`  1. Edit ${result.targetDir}/SKILL.md (review fields)\n`);
    process.stdout.write(`  2. windy-drops validate ${result.targetDir}\n`);
    process.stdout.write(`  3. windy-drops publish ${result.targetDir}\n`);
    process.exit(0);
  } catch (e) {
    process.stderr.write(`${chalk.red("✗")} ${(e as Error).message}\n`);
    process.exit(1);
  }
}

File: python/sdk/src/windy_drops/commands/fork.py
"""fork command — WD-10. Locally fork a drop: clone the bundle source, rewrite
SKILL.md (id, author, forked_from, version reset to 1.0.0).

In v1 we accept a LOCAL source via --from. Once the registry ships
(WD-12+), this command will also accept a registry drop-id and fetch
the bundle automatically (plus POST /drops/{id}/fork — WD-19).
"""

from __future__ import annotations

import re
import shutil
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import typer
import yaml

from ..lib.eternitas import load_credential

FRONTMATTER_RE = re.compile(
    r"^---\r?\n(?P<yaml>[\s\S]*?)\r?\n---(?P<body>\r?\n[\s\S]*)?$"
)


@dataclass(frozen=True)
class ForkResult:
    source_dir: Path
    target_dir: Path
    new_id: str
    forked_from: str


def fork(
    *,
    from_: str,
    new_id: str,
    out: str,
    new_name: str | None = None,
    author_name: str | None = None,
    author_passport: str | None = None,
) -> ForkResult:
    source = Path(from_).resolve()
    if not source.exists():
        raise FileNotFoundError(f"source not found: {source}")

    target = Path(out).resolve()
    if target.exists():
        if any(target.iterdir()):
            raise FileExistsError(f"target directory not empty: {target}")
    else:
        target.mkdir(parents=True)

    for entry in source.iterdir():
        if entry.is_dir():
            shutil.copytree(entry, target / entry.name)
        else:
            shutil.copy2(entry, target / entry.name)

    skill_path = target / "SKILL.md"
    raw = skill_path.read_text(encoding="utf-8")
    match = FRONTMATTER_RE.match(raw)
    if not match:
        raise ValueError(f"source SKILL.md has no frontmatter: {skill_path}")

    manifest: dict[str, Any] = yaml.safe_load(match.group("yaml") or "") or {}
    if not isinstance(manifest, dict):
        raise ValueError(f"source SKILL.md frontmatter must be a YAML object: {skill_path}")

    source_id = str(manifest.get("id", ""))
    source_name = manifest.get("name")

    if author_name:
        author: dict[str, Any] = {"name": author_name}
        if author_passport:
            author["passport"] = author_passport
        new_author = [author]
    else:
        try:
            cred = load_credential()
            new_author = [
                {
                    "name": cred.passport or "Your Name",
                    **({"passport": cred.passport} if cred.passport else {}),
                }
            ]
        except (FileNotFoundError, ValueError):
            new_author = [{"name": "Your Name"}]

    manifest["id"] = new_id
    manifest["author"] = new_author
    manifest["forked_from"] = source_id
    manifest["version"] = "1.0.0"
    if new_name:
        manifest["name"] = new_name
    elif isinstance(source_name, str):
        manifest["name"] = f"{source_name} (forked)"
    manifest.pop("signature", None)

    yaml_out = yaml.safe_dump(
        manifest, sort_keys=False, default_flow_style=False, allow_unicode=True, width=10_000
    ).rstrip()
    body = match.group("body") or ""
    skill_path.write_text(f"---\n{yaml_out}\n---{body}", encoding="utf-8")

    return ForkResult(
        source_dir=source,
        target_dir=target,
        new_id=new_id,
        forked_from=source_id,
    )


def run(
    new_id: str = typer.Argument(..., help="ID for your fork"),
    target: str = typer.Argument(..., help="Directory to write the fork into (must be empty)"),
    from_: str = typer.Option(..., "--from", help="Local source drop directory"),
    name: str | None = typer.Option(None, "--name", help="Human-readable name"),
    author_name: str | None = typer.Option(None, "--author-name", help="Explicit author name"),
    author_passport: str | None = typer.Option(None, "--author-passport", help="Explicit author passport"),
) -> None:
    """Clone a drop locally and rewrite its manifest."""
    try:
        result = fork(
            from_=from_,
            new_id=new_id,
            out=target,
            new_name=name,
            author_name=author_name,
            author_passport=author_passport,
        )
    except (FileNotFoundError, FileExistsError, ValueError) as e:
        typer.echo(f"✗ {e}", err=True)
        sys.exit(1)
    typer.echo(f"✓ forked {result.forked_from} → {result.new_id}")
    typer.echo(f"  at {result.target_dir}\n")
    typer.echo("Next steps:")
    typer.echo(f"  1. Edit {result.target_dir}/SKILL.md (review fields)")
    typer.echo(f"  2. windy-drops validate {result.target_dir}")
    typer.echo(f"  3. windy-drops publish {result.target_dir}")
